src/solver/main.jl

```diff
@@ -0,0 +1,293 @@
+The base solver (`symbolic_solve`) has multiple solvers which chooses from depending on the the type of input
+(multiple/uni var and multiple/single expression) only after ensuring that the input is valid.
+
+The expressions inputted can contain parameters, which are assumed to be transcendental.
+A parameter "a" is transcendental if there exists no polynomial P with
+rational coefficients such that P(a) = 0. Check the examples section.
+
+Currently, `symbolic_solve` supports
+- Linear and polynomial equations (with parameters)
+- Systems of linear and polynomials equations (without extra parameters, for now)
+- Equations with transcendental functions (with parameters)
+
+## Examples
+
+TODO(Alex): add a parametric polynomial solve (perhaps smaller) to show-case `roots_of`:
+julia> f = expand((x - a)^2 * (x^10 - 1) * (a*x^2 + b*x * c) * (x^20 + a))
+julia> RootFinding.solve(f, x)
+
+### `solve_univar` (uses factoring and analytic solutions up to degree 4)
+```jldoctest
+julia> expr = expand((x + b)*(x^2 + 2x + 1)*(x^2 - a))
+-a*b - a*x - 2a*b*x - 2a*(x^2) + b*(x^2) + x^3 - a*b*(x^2) - a*(x^3) + 2b*(x^3) + 2(x^4) + b*(x^4) + x^5
+
+julia> Symbolics.symbolic_solve(expr, x)
+4-element Vector{Any}:
+ -1
+   -b
+   (1//2)*√(4a)
+   (-1//2)*√(4a)
+
+julia> Symbolics.symbolic_solve(expr, x, dropmultiplicity=false)
+5-element Vector{Any}:
+ -1
+ -1
+   -b
+   (1//2)*√(4a)
+   (-1//2)*√(4a)
+```
+```jldoctest
+julia> Symbolics.symbolic_solve(x^2 + a*x + 6, x)
+2-element Vector{SymbolicUtils.BasicSymbolic{Real}}:
+ (1//2)*(-a + √(-24 + a^2))
+ (1//2)*(-a - √(-24 + a^2))
+```
+```jldoctest
+julia> symbolic_solve(x^7 - 1, x)
+2-element Vector{Any}:
+  roots_of((1//1) + x + x^2 + x^3 + x^4 + x^5 + x^6, x)
+ 1
+```
+### `solve_multivar` (uses Groebner basis and `solve_univar` to find roots)
+```jldoctest
+julia> using Groebner
+
+julia> @variables x y z
+3-element Vector{Num}:
+ x
+ y
+ z
+
+julia> eqs = [x+y^2+z, z*x*y, z+3x+y]
+3-element Vector{Num}:
+ x + z + y^2
+       x*y*z
+  3x + y + z
+
+julia> symbolic_solve(eqs, [x,y,z])
+3-element Vector{Any}:
+ Dict{Num, Any}(z => 0, y => 1//3, x => -1//9)
+ Dict{Num, Any}(z => 0, y => 0, x => 0)
+ Dict{Num, Any}(z => -1, y => 1, x => 0)
+```
+
+
+### `solve_multipoly` (uses GCD between the input polys)
+```jldoctest
+julia> symbolic_solve([x-1, x^3 - 1, x^2 - 1, (x-1)^20], x)
+1-element Vector{BigInt}:
+ 1
+```
+
+### `ia_solve` (solving by isolation and attraction)
+```jldoctest
+julia> symbolic_solve(2^(x+1) + 5^(x+3), x)
+1-element Vector{SymbolicUtils.BasicSymbolic{Real}}:
+ (-slog(2) - log(complex(-1)) + 3slog(5)) / (slog(2) - slog(5))
+```
+```jldoctest
+julia> symbolic_solve(log(x+1)+log(x-1), x)
+2-element Vector{SymbolicUtils.BasicSymbolic{BigFloat}}:
+ (1//2)*√(8.0)
+ (-1//2)*√(8.0)
+```
+```jldoctest
+julia> symbolic_solve(a*x^b + c, x)
+((-c)^(1 / b)) / (a^(1 / b))
+```
+"""
+function symbolic_solve(expr, x; dropmultiplicity=true)
```

Comment:
option to disable throws?

Reply:
sure, will work on it, it should obviously warn the user and return nothing right?

Reply:
> option to disable throws?

That's a good point.

Some parts of the solver throw a string with info, for example,

```julia
if length(new_eqs) < length(vars)
        throw("Infinite number of solutions")
end
```

or

```julia
       isequal(old_lhs, lhs) && throw("symbolic_solve can not currently solve this expression")
```

Instead of throwing it can return an empty thing (array, dict, ...). Optionally, also print the message.

This can be the default. Not sure if this needs a separate keyword argument.

Reply:
asserts must throw though, right ? @ChrisRackauckas

Reply:
Assertions yes. So things like malformed inputs should throw, but something like didn't solve should instead have a code.